{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Samantha Hudson, Zak Martinjako, Jeremy Pigat, David Ramsay\n",
    "\n",
    "All cells must be run on first starting up notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#difference matrix\n",
    "#Both sequences must be aligned\n",
    "def getDistance(sequence,sequence2):\n",
    "    distance=0\n",
    "    i=0\n",
    "    while i < len(sequence):\n",
    "        if sequence[i]!=sequence2[i]:\n",
    "            distance+=1\n",
    "        i+=1\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Node:\n",
    "    def __init__(self,name='',distance=0,leftChild=None,rightChild=None,height=0):\n",
    "        self.name=name\n",
    "        self.distance=distance\n",
    "        self.leftChild=leftChild\n",
    "        self.rightChild=rightChild\n",
    "        self.height=height\n",
    "    def countNodes(self):\n",
    "        if(self.leftChild==None) and (self.rightChild==None):\n",
    "            return 1\n",
    "        else:\n",
    "            return 1 + self.leftChild.countNodes() + self.rightChild.countNodes()\n",
    "    def setName(self,newName):\n",
    "        self.name=newName\n",
    "    def setLeft(self,left):\n",
    "        self.leftChild=left\n",
    "    def setRight(self,right):\n",
    "        self.rightChild=right\n",
    "    def setDistance(self,dist):\n",
    "        self.distance=dist\n",
    "    def setHeight(self,height):\n",
    "        self.height=height\n",
    "    def getDistance(self):\n",
    "        return self.distance\n",
    "    def __str__(self):\n",
    "        return str(self.name)\n",
    "    def printNode(self):\n",
    "        #For valid Newick format that TreeDyn will accept:\n",
    "        #-cannot use the names of any non-leaf nodes, since they are tuples and tuples break the formatting\n",
    "        #---due to parenthesis. Can get around this by only printing the names of leaf nodes\n",
    "        #-cannot have whitespace\n",
    "        #---replace whitespace with _\n",
    "        #base case\n",
    "        if self.leftChild==None and self.rightChild==None:\n",
    "            name = self.name\n",
    "            if(' ' in name):\n",
    "                name = \"_\".join(name.split())\n",
    "            return name+\":\"+str(self.distance)\n",
    "        out=\"(\"+self.leftChild.printNode()+\",\"+self.rightChild.printNode()+\")\"+\\\n",
    "        (\"\" if self.distance==0 else \":\"+str(self.distance))\n",
    "        return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createGeneList(filepath):\n",
    "    \"\"\"\n",
    "    Given the filepath to a FASTA formatted file which contains a genome sequence divided into each gene, creates\n",
    "    a dictionary where each header line is a key and the value to each key is the sequence of the gene which has been\n",
    "    stripped of whitespace. Returns a dictionary of this format -> header_line(str):gene_sequence(str).\n",
    "    \n",
    "    An incorrect filepath input will result in a return of None.\n",
    "    \"\"\"\n",
    "    if(type(filepath) is not str):\n",
    "        print(\"ERROR1\")\n",
    "        return\n",
    "    try:\n",
    "        fh = open(filepath)\n",
    "    except IOError:\n",
    "        print(\"ERROR2\")\n",
    "        return\n",
    "    gene_list = []\n",
    "    current_name = \"\"\n",
    "    current_seq = \"\"\n",
    "    for line in fh.readlines():\n",
    "        if(line.startswith(\">\")):\n",
    "            if((current_name != '') and (current_name is not None) and (current_seq != '') and (current_seq is not None)):\n",
    "                #if a > line is reached, and there are still values for current name and seq, that gene has ended and\n",
    "                #should be added to the list.\n",
    "                gene_list.append([current_name, current_seq])\n",
    "                current_name = \"\"\n",
    "                current_seq = \"\"\n",
    "            line = line.replace('>','')\n",
    "            line = line.replace('\\n','')\n",
    "            current_name = line\n",
    "        elif(line.startswith(\"\\n\")): #if the line starts with a newline, this indicates the end of the gene\n",
    "            gene_list.append([current_name, current_seq])\n",
    "            current_name = \"\"\n",
    "            current_seq = \"\"\n",
    "        else:\n",
    "            #every line with base information is concatenated with current_value\n",
    "            line = line.replace('\\n','')\n",
    "            current_seq = current_seq + line\n",
    "    if((current_name != '') and (current_name is not None) and (current_seq != '') and (current_seq is not None)):\n",
    "        #handles the case if there was no newline character after the last gene in the file.\n",
    "        gene_list.append([current_name, current_seq])\n",
    "        current_name = \"\"\n",
    "        current_seq = \"\"\n",
    "    fh.close()\n",
    "    return gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def buildMatrix(filepath):\n",
    "    distanceMatrix={}\n",
    "    sequences=[]\n",
    "    seqs = createGeneList(filepath) \n",
    "    for i in range(0,len(seqs)):\n",
    "        sequences.append(seqs[i][0])\n",
    "        for j in range(i+1,len(seqs)):\n",
    "            distanceMatrix[(seqs[i][0],seqs[j][0])]=getDistance(seqs[i][1],seqs[j][1])\n",
    "    return (distanceMatrix,sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def UPGMA(distMatrix, seqs):\n",
    "    #Create list of leaf nodes\n",
    "    nodeList=[]\n",
    "    for seq in seqs:\n",
    "        nodeList.append(Node(seq))\n",
    "        \n",
    "    #UPGMA Algorithm\n",
    "    while len(distMatrix) > 2:\n",
    "        #Find the minimum distance in matrix, get the names of nodes that have this distance and the value of the distance\n",
    "        minDist = min(distMatrix, key=distMatrix.get)\n",
    "        distVal = distMatrix[minDist]\n",
    "        seq1, seq2 = minDist\n",
    "        \n",
    "        #Using the node names from the matrix, retrieve the node objects from the list of nodes.\n",
    "        #The nodes you aren't looking for are added to a new nodeList, which will be used further down\n",
    "        #which will essentially \"delete\" the nodes we are about to use from the list of nodes.\n",
    "        node1=None\n",
    "        node2=None\n",
    "        newNodeList=[]\n",
    "        \n",
    "        for node in nodeList:\n",
    "            if node.name==seq1:\n",
    "                node1=node\n",
    "                continue\n",
    "            if node.name==seq2:\n",
    "                node2=node\n",
    "                continue\n",
    "            newNodeList.append(node)\n",
    "        \n",
    "        #Set the distance of the nodes we retrieved, as it wasn't set at their creation (since it was not yet known)\n",
    "        node1.setDistance(distVal/2-node1.height)\n",
    "        node2.setDistance(distVal/2-node2.height)\n",
    "        \n",
    "        #Create a new node that will form a subtree, with node1 and node2 as the left and right children\n",
    "        clusterNode=Node(\"(\"+seq1+\",\"+seq2+\")\", 0, node1, node2, distVal/2)\n",
    "        \n",
    "        #Set nodeList to be the new nodeList, effectively removing node1 and node2 from the nodeList\n",
    "        nodeList=newNodeList\n",
    "        \n",
    "        #Re-calculate the distance matrix\n",
    "        for i in nodeList:\n",
    "            #Create the key for the distMatrix. I.E. if \n",
    "            newNodeName = (\"(\"+node1.name+\",\"+node2.name+\")\",i.name)\n",
    "            \n",
    "            #Get the number of nodes in each branch of which the new node's children are the root\n",
    "            n1count = node1.countNodes()\n",
    "            n2count = node2.countNodes()\n",
    "            \n",
    "            #Find the value in the distMatrix\n",
    "            #Have to check each possible combination of node1.name and i.name, and node2.name and i.name\n",
    "            if(((node1.name, i.name) in distMatrix) and ((node2.name, i.name) in distMatrix)):\n",
    "                distN1i = distMatrix[node1.name,i.name]\n",
    "                distN2i = distMatrix[node2.name,i.name]\n",
    "            elif(((i.name, node1.name) in distMatrix) and ((i.name, node2.name) in distMatrix)):\n",
    "                distN1i = distMatrix[i.name, node1.name]\n",
    "                distN2i = distMatrix[i.name, node2.name]\n",
    "            elif(((i.name, node1.name) in distMatrix) and ((node2.name, i.name) in distMatrix)):\n",
    "                distN1i = distMatrix[i.name, node1.name]\n",
    "                distN2i = distMatrix[node2.name, i.name]\n",
    "            elif(((node1.name, i.name) in distMatrix) and ((i.name, node2.name) in distMatrix)):\n",
    "                distN1i = distMatrix[node1.name, i.name]\n",
    "                distN2i = distMatrix[i.name, node2.name]\n",
    "            \n",
    "            #Calculate the distance for the entry in the distMatrix\n",
    "            distMatrix[newNodeName] = ((distN1i*n1count)+(distN2i*n2count))/(n1count+n2count)        \n",
    "            \n",
    "            #distMatrix[newNodeName]=(distN1i+distN2i)/2 --- Way we originally calculated distance\n",
    "            #--Usually gives the same result as the way above, but the slides use the above equation \n",
    "            #--rather than this one, so I assume that there is some case where it matters, so I am using \n",
    "            #--the equation from the slides.\n",
    "        \n",
    "        #Add the newly made clusterNode to the nodeList\n",
    "        nodeList.append(clusterNode)\n",
    "        \n",
    "        #Find all entries in distMatrix which directly reference node1 and node2 and remove it\n",
    "        deleteList=[]\n",
    "        for first,second in distMatrix:\n",
    "            if first==node1.name or first==node2.name or second==node1.name or second==node2.name:\n",
    "                deleteList.append((first,second))\n",
    "        for d in deleteList:\n",
    "            del distMatrix[d]\n",
    "\n",
    "    #Combine final 2 nodes\n",
    "    distValue=0\n",
    "    for d in distMatrix: #there should only be 1 value for d\n",
    "        distValue+=distMatrix[d]\n",
    "    distValue=distValue\n",
    "    root = Node(\"(\"+nodeList[0].name+\",\"+nodeList[1].name+\")\", 0, nodeList[0], nodeList[1], distValue/2)\n",
    "    #root.setDistance((nodeList[0].distance + nodeList[1].distance)/2) --- not needed, the root has no distance\n",
    "    root.leftChild.distance=distValue/2-root.leftChild.height\n",
    "    root.rightChild.distance=distValue/2-root.rightChild.height\n",
    "    \n",
    "    #Return Newick formatted tree\n",
    "    return root.printNode()+\";\" #need to add ; at the end for TreeDyn to recognized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Newick:\n",
      "(15Quercuspetraea:639.5327747889927,(((((((10pindrowAfghanistantoNepal:10.5,12chensiensisTibetIndiaChina:10.5):5.25,9kawakamiiTaiwan:15.75):3.75,(6forrestiiChina:10.0,7delavayiSouthernChina:10.0):9.5):7.6640625,5cephalonicaGreekMountains:27.1640625):3.7003906250000007,((13ziyuanensisChina:5.5,14fargesiiCentralChina:5.5):1.0,11fanjingshanensisFanjingMountainChina:6.5):24.364453125):10.930615234375,((3amabilisPacificCoast:9.0,4mariesiiNorthernJapan:9.0):7.25,(0proceraWestUSACanada:4.0,1magnificaSouthwestOregonCalifornia:4.0):12.25):25.545068359375):12.798746744791671,(2concolorSouthernOregon:40.0,8fabriWesternChina:40.0):14.593815104166673):584.938959684826);\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mat,seqs=buildMatrix(\"msa.txt\")\n",
    "print(\"Newick:\\n\"+UPGMA(mat,seqs)+\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
