{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Running this algorithm is simple, simply run all of the cells in order. \n",
    "The program is setup to run a text file called 'msa.txt' and you can\n",
    "change that by changing the name of the file in the last cell.\n",
    "\n",
    "The expected output of this program is as follows:\n",
    "\n",
    "\n",
    "\n",
    "(15Quercuspetraea:639.5327747889927,(((((((10pindrowAfghanistantoNepal:10.5,\n",
    "12chensiensisTibetIndiaChina:10.5):5.25,9kawakamiiTaiwan:15.75):3.75,\n",
    "(6forrestiiChina:10.0,7delavayiSouthernChina:10.0):9.5):7.6640625,\n",
    "5cephalonicaGreekMountains:27.1640625):3.7003906250000007,\n",
    "((13ziyuanensisChina:5.5,14fargesiiCentralChina:5.5):1.0,\n",
    "11fanjingshanensisFanjingMountainChina:6.5):24.364453125):10.930615234375,\n",
    "((3amabilisPacificCoast:9.0,4mariesiiNorthernJapan:9.0):7.25,\n",
    "(0proceraWestUSACanada:4.0,1magnificaSouthwestOregonCalifornia:4.0):12.25):25.545068359375):12.798746744791671,\n",
    "(2concolorSouthernOregon:40.0,8fabriWesternChina:40.0):14.593815104166673):584.938959684826);\n",
    "\n",
    "The output above is what this algorithm output.\n",
    "\n",
    "It is different than when I put the msa.txt file into an online tree maker,\n",
    "but I believe that is due to the different algorithms. The online tool\n",
    "I used output a Neighbor-Joining tree, and I could not find a UPGMA tree that\n",
    "accepted my genes, since I changed the names to show location.\n",
    "\n",
    "Many of them only accepted the official NCBI species names, so I could\n",
    "not get them to work without recollecting all of my data.\n",
    "\n",
    "Since this outputs in Newick format, the tree is not built for you.\n",
    "\n",
    "I use the tool made by Interactive Tree of life and that can be found\n",
    "here: https://itol.embl.de\n",
    "\n",
    "My specific tree can be found here: \n",
    "https://itol.embl.de/tree/14320619210946931523500649\n",
    "\n",
    "This link should work, but the website most likely won't host it forever.\n",
    "\n",
    "If I have time, I will implement that, but as of 09 April 2018, I have\n",
    "yet to do so.\n",
    "\n",
    "This program was written by William Wright\n",
    "Hiram username: WrightWJ\n",
    "Final Date: 04-11-2018 11:36 PM.\n",
    "\n",
    "\n",
    "Note: Thank you for the class, Louis. I always enjoy your classes.\n",
    "\n",
    "\"\"\"\n",
    "def getDist(seq1,seq2):\n",
    "    distance=0\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            distance+=1\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Louis made this in class, I copied what I could and made slight modifications\n",
    "# to promote ease of use, like the distance perameter. \n",
    "class Node:\n",
    "    def __init__(self,name='',distance=0,left=None,right=None,height=0):\n",
    "        self.name=name\n",
    "        self.distance=distance\n",
    "        self.left=left\n",
    "        self.right=right\n",
    "        self.height=height\n",
    "    def countChildren(self):\n",
    "        if(self.left==None) and (self.right==None):\n",
    "            return 1\n",
    "        else:\n",
    "            return 1 + self.left.countChildren() + self.right.countChildren()\n",
    "    def setName(self,newName):\n",
    "        self.name=newName\n",
    "    def setLeft(self,left):\n",
    "        self.left=left\n",
    "    def setRight(self,right):\n",
    "        self.right=right\n",
    "    def setDistance(self,dist):\n",
    "        self.distance=dist\n",
    "    def setHeight(self,height):\n",
    "        self.height=height\n",
    "    def getDistance(self):\n",
    "        return self.distance\n",
    "    def __str__(self):\n",
    "        return str(self.name)\n",
    "    def printNode(self):\n",
    "        if self.left==None and self.right==None:\n",
    "            return self.name+\":\"+str(self.distance)\n",
    "        out=\"(\"+self.left.printNode()+\",\"+self.right.printNode()+\")\"+\\\n",
    "        (\"\" if self.distance==0 else \":\"+str(self.distance))\n",
    "        return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createGenes(file):\n",
    "    # This imports file, sorts data into a dictionary where the key is the name, the value is the \n",
    "    # sequence.\n",
    "    # There must be a newline after each gene.\n",
    "    # File name must be a string, and the file of Multiple Sequence Alignments in FASTA format\n",
    "    # must be in the same directory\n",
    "    fh = open(file)\n",
    "    data=fh.readlines()\n",
    "    fh.close()\n",
    "    # entire list of genes here\n",
    "    genes=[]\n",
    "    \n",
    "    #for storage while sorting\n",
    "    name=\"\"\n",
    "    seq=\"\"\n",
    "\n",
    "    for line in data:\n",
    "        #if the line is a new gene\n",
    "        if (line.startswith(\">\")):\n",
    "            #if the marker is now on a new gene, append the previous value\n",
    "            if ((name != '') and (name is not None) and (seq != '') and (seq is not None)):\n",
    "                genes.append([name,seq])\n",
    "                name=\"\"\n",
    "                seq=\"\"\n",
    "            #delete > at front so the name is only the name\n",
    "            line = line.replace('>','')\n",
    "            line = line.replace('\\n','')\n",
    "            #delete any newlines so the genetic data is the only thing in the sequence\n",
    "            name=line\n",
    "        # Appends gene once newline is found\n",
    "        elif (line.startswith(\"\\n\")):\n",
    "            genes.append([name,seq])\n",
    "            seq=\"\"\n",
    "            name=\"\"\n",
    "        else:\n",
    "            line = line.replace('\\n','')\n",
    "            seq=seq+line\n",
    "    # This should only be run if the file is formatted incorrectly\n",
    "    if((name!='') and (name is not None) and (seq is not None) and (seq!='')):\n",
    "        genes.append([name,seq])\n",
    "        name=\"\"\n",
    "        seq=\"\"\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix(file):\n",
    "    #this stores the distance matrix for further upgma use.\n",
    "    beginMatrix={}\n",
    "    \n",
    "    #stores sequence names\n",
    "    seqs=[]\n",
    "    \n",
    "    #get the list of genes for the beginning of the distance matrix.\n",
    "    genes = createGenes(file)\n",
    "    \n",
    "    #begin creation\n",
    "    for i in range(0,len(genes)):\n",
    "        seqs.append(genes[i][0])\n",
    "        #i + 1 because it should only compare each gene once, not every gene to itself and \n",
    "        # others twice.\n",
    "        for j in range(i+1,len(genes)):\n",
    "            #gets distance of the two sequences and stores in in a matrix with a tuple as the key\n",
    "            beginMatrix[(genes[i][0],genes[j][0])] = getDist(genes[i][1],genes[j][1])\n",
    "            \n",
    "    return beginMatrix,seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def UPGMA(matrix,seqs):\n",
    "    nodes=[]\n",
    "    #Makes a list of all leaf nodes in the matrix using the names of the sequences\n",
    "    for seq in seqs:\n",
    "        nodes.append(Node(seq))\n",
    "        \n",
    "    #Begin Algorithm\n",
    "    while len(matrix)>2:\n",
    "        #get min distance and index, which is the key and value in matrix\n",
    "        newMin=min(matrix,key=matrix.get)\n",
    "        calcDist = matrix[newMin]\n",
    "        seq1,seq2 = newMin\n",
    "        \n",
    "        #get nodes\n",
    "        #the nodes that don't contain the minimum\n",
    "        #will be added to the notUsed list\n",
    "        node1=None\n",
    "        node2=None\n",
    "        notUsed=[]\n",
    "        \n",
    "        #add nodes to list to delete later\n",
    "        for node in nodes:\n",
    "            if node.name==seq1:\n",
    "                node1=node\n",
    "                continue\n",
    "            if node.name==seq2:\n",
    "                node2=node\n",
    "                continue\n",
    "            notUsed.append(node)\n",
    "            \n",
    "        #set heights of new nodes\n",
    "        node1.setDistance((calcDist/2) - node1.height)\n",
    "        node2.setDistance((calcDist/2) - node2.height)\n",
    "        \n",
    "        #new node with children node 1 and node 2\n",
    "        subTree = Node(\"(\"+seq1+\",\"+seq2+\")\",0,node1,node2,calcDist/2)\n",
    "        \n",
    "        #nodes now exclude previous nodes, but include the subTree node\n",
    "        nodes = notUsed\n",
    "        \n",
    "        #New Matrix\n",
    "        for node in nodes:\n",
    "            #newNode is the key for the new distance matrix\n",
    "            subTreeName = (\"(\"+node1.name+\",\"+node2.name+\")\",node.name)\n",
    "            #gets a height of sorts; n1 is node 1's count of nodes including\n",
    "            #itself and all its children\n",
    "            n1 = node1.countChildren()\n",
    "            n2 = node2.countChildren()\n",
    "            \n",
    "            # check each combination and find new distance value\n",
    "            # node1 is left child, node2 is right, and node is current node in list\n",
    "            # I named them node, node1, and node2 for ease of copy and paste.\n",
    "            #\n",
    "            #\n",
    "            # This is really messy, but I'm not sure how else to do it.\n",
    "            # Even other people I talked to during group project 3 had this\n",
    "            if (((node1.name,node.name) in matrix) and ((node2.name,node.name) in matrix)): \n",
    "                distn1 = matrix[node1.name,node.name]\n",
    "                distn2 = matrix[node2.name,node.name]\n",
    "            elif (((node.name,node1.name) in matrix) and ((node.name,node2.name) in matrix)):\n",
    "                distn1 = matrix[node.name,node1.name]\n",
    "                distn2 = matrix[node.name,node2.name]\n",
    "            elif (((node.name,node1.name) in matrix) and ((node2.name,node.name) in matrix)):\n",
    "                distn1 = matrix[node.name,node1.name]\n",
    "                distn2 = matrix[node2.name,node.name]\n",
    "            elif(((node1.name,node.name) in matrix) and ((node.name,node2.name) in matrix)):\n",
    "                distn1 = matrix[node1.name,node.name]\n",
    "                distn2 = matrix[node.name,node2.name]\n",
    "            # get new distance\n",
    "            # After my presentation, I changed it to division based on the number of nodes.\n",
    "            # It's usually the same, but in the case of a 3 way tie (pretty rare) this will be more accurate\n",
    "            # The old method divided by two, but was incorrect. I believe that was where my error was.\n",
    "            matrix[subTreeName]=((distn1*n1)+(distn2*n2))/(n1+n2)\n",
    "        # Finally, append the node to the overall matrix. \n",
    "        nodes.append(subTree)\n",
    "        \n",
    "        #delete all nodes from matrix with node1 or node2 in it\n",
    "        nodesToDelete=[]\n",
    "        for name,val in matrix:\n",
    "            if name==node1.name or name==node2.name or val == node1.name or val == node2.name:\n",
    "                nodesToDelete.append((name,val))\n",
    "                \n",
    "        for node in nodesToDelete:\n",
    "            del matrix[node]\n",
    "            \n",
    "    # combine the last two then done!\n",
    "    # easier said than implemented\n",
    "        \n",
    "    # get last distance\n",
    "    \"\"\"\n",
    "    Funny side note for Louis or whoever is reading this:\n",
    "    I could not figure out why my code was working for a couple of weeks\n",
    "    I tried everything. I asked my friends if they could see anything wrong with it\n",
    "    (a lot of them aren't in bioinformatics, so they had no clue)\n",
    "    No one could tell me what the error was.\n",
    "    \n",
    "    I should have gone to you, Louis, for help, right? See, I knew I was close\n",
    "    to figuring it out, I just couldn't tell why I was wrong.\n",
    "    \n",
    "    Do you want to know what was wrong about this little bit of code below?\n",
    "    \n",
    "    Well. It should have been\n",
    "    here.\n",
    "    Instead, it was aligned\n",
    "        here. And Python does not like misaligned code. It's amazing why I didn't\n",
    "    see it earlier, but I thought it was hilarious when I discovered it.\n",
    "    I hope you enjoyed this text. I figure a bit of humor will be nice when you\n",
    "    are grading a bunch of IRCs.\"\"\"\n",
    "    distUpdate=0\n",
    "    for dist in matrix:\n",
    "        distUpdate+=matrix[dist]\n",
    "    # set height of root node from children\n",
    "    distFromKids=distUpdate/2\n",
    "        \n",
    "    # create root node with no distance since it is root\n",
    "    root=Node(\"(\"+nodes[0].name+\",\"+nodes[1].name+\")\",0, nodes[0], nodes[1],distFromKids)\n",
    "    # set children to correct nodes\n",
    "    root.left.distance=distFromKids-root.left.height\n",
    "    root.right.distance=distFromKids-root.right.height\n",
    "        \n",
    "    #return tree that should be in Newick format.\n",
    "    return root.printNode()+\";\" #for TreeDyn, which was used to visualize the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Newick:\n",
      "(15Quercuspetraea:639.5327747889927,(((((((10pindrowAfghanistantoNepal:10.5,12chensiensisTibetIndiaChina:10.5):5.25,9kawakamiiTaiwan:15.75):3.75,(6forrestiiChina:10.0,7delavayiSouthernChina:10.0):9.5):7.6640625,5cephalonicaGreekMountains:27.1640625):3.7003906250000007,((13ziyuanensisChina:5.5,14fargesiiCentralChina:5.5):1.0,11fanjingshanensisFanjingMountainChina:6.5):24.364453125):10.930615234375,((3amabilisPacificCoast:9.0,4mariesiiNorthernJapan:9.0):7.25,(0proceraWestUSACanada:4.0,1magnificaSouthwestOregonCalifornia:4.0):12.25):25.545068359375):12.798746744791671,(2concolorSouthernOregon:40.0,8fabriWesternChina:40.0):14.593815104166673):584.938959684826);\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix0,seqs=matrix(\"msa.txt\")\n",
    "print(\"Newick:\\n\"+UPGMA(matrix0,seqs)+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
